external-crates/move/crates/move-compiler/src/naming/syntax_methods.rs

```diff
@@ -0,0 +1,514 @@
+                            Some(type_name.clone())
+                        } else {
+                            context.env.add_diag(diag!(
+                                Declarations::InvalidSyntaxMethod,
+                                (*ann_loc, INVALID_MODULE_MSG),
+                                (*loc, INVALID_MODULE_TYPE_MSG)
+                            ));
+                            None
+                        }
+                    } else {
+                        context.env.add_diag(diag!(
+                            Declarations::InvalidSyntaxMethod,
+                            (*ann_loc, "The subject type is a Move builtin type"),
+                        ));
+                        None
+                    }
+                } // something like fake_natives?
+                N::TypeName_::ModuleType(defining_module, _n) => {
+                    if cur_module == defining_module {
+                        Some(type_name.clone())
+                    } else {
+                        context.env.add_diag(diag!(
+                            Declarations::InvalidSyntaxMethod,
+                            (*ann_loc, INVALID_MODULE_MSG),
+                            (*loc, INVALID_MODULE_TYPE_MSG)
+                        ));
+                        None
+                    }
+                }
+            }
+        }
+        N::Type_::Ref(_, inner) => determine_subject_type_name(context, cur_module, ann_loc, inner),
+        N::Type_::Param(param) => {
+            let msg = format!(
+                "Invalid {} annotation. Cannot associate a syntax method with a type parameter",
+                SyntaxAttribute::SYNTAX
+            );
+            let tmsg = format!(
+                "But '{}' was declared as a type parameter here",
+                param.user_specified_name
+            );
+            context.env.add_diag(diag!(
+                Declarations::InvalidSyntaxMethod,
+                (*ann_loc, msg),
+                (*loc, tmsg)
+            ));
+            None
+        }
+        N::Type_::Var(_) | N::Type_::Anything | N::Type_::UnresolvedError => {
+            assert!(context.env.has_errors());
+            None
+        }
+        N::Type_::Unit | N::Type_::Fun(_, _) => {
+            let msg = "Invalid type for syntax method definition";
+            let mut diag = diag!(Declarations::InvalidSyntaxMethod, (*loc, msg));
+            diag.add_note("Syntax methods may only be defined for single base types");
+            context.env.add_diag(diag);
+            None
+        }
+    }
+}
+
+fn valid_return_type(
+    context: &mut Context,
+    sp!(loc, kind_): &SyntaxMethodKind,
+    ty: &N::Type,
+) -> bool {
+    match kind_ {
+        SyntaxMethodKind_::Index if valid_imm_ref(ty) => valid_index_return_type(context, loc, ty),
+        SyntaxMethodKind_::Index if valid_mut_ref(ty) => {
+            let msg = format!("Invalid {} annotation", SyntaxAttribute::SYNTAX);
```

Comment:
Normally love guards, but it might just be easier to match inside the kind here

Reply:
my days of using `cond` are showing